### PR TITLE
Add support for Docker default config file

### LINKED ISSUE #606: Feature Request: Read docker credentials from ~/.docker/config.json
### Is your feature request related to a problem?

At the moment you need to specify the username/passsword of a docker registry in the manifest file.
This does make it tricky to check into version control, or other wise manage.

The docker cli and tooling, do already configure some of this information in ~/.docker/config.json. So most people would have this already configured.

### Solution you'd like

Rather than specify in the manifest. Check the config.json and use credentials specified there on registries listed.
This would make it easy to use the tool, as if you have logged using docker cli. The application should get the right auth.

### Alternatives you've considered

_No response_

### Anything else?

Split out from #600 as it is it's own feature.

### LINKED ISSUE #871: dockerimage authentioncation issues.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When providing a `token` to the dockerimage it doesn't use the provided token. 

Additionally, for docker registries like artifactory, the default username/password mechanism used by updatecli doesn't seem to work either in authenticated mode or in anonymous mode.

```
DEBUG: Searching tags for the image "registry.io/namespace/name"
DEBUG: Retrieving a bearer token from https://registry.io/token?scope=repository:namespace:pull&service=registry.io with the username  and the associated password.
WARNING: No username and/or password specified: trying to retrieve a token as anonymous user might not be supported or could impact the rate limiting on your network.
DEBUG: Emitting a request to https://registry.io/token?scope=repository:namespace:pull&service=registry.io
ERROR: Got an HTTP response code 404 while logging in to the registry registry.io.
ERROR: ✗ Got an HTTP response code 404 while logging in to the registry registry.io.
```

### Expected Behavior

That the above works.

### Steps To Reproduce

Create a manifest like so - 

```yaml
sources:
  test:
    kind: dockerimage
    spec:
      image: myregistry.io/test/test
      token: 'test'
```

### Environment

```markdown
- OS: macos
- updatecli: 0.33.2
```

### Anything else?

_No response_

### LINKED ISSUE #872: parsing issues with nested container images
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Currently updatecli throws warning like

> WARNING: No architecture specified for the image "myregistry.io/nested/namespace:". Using the default "amd64".

if an image like `myregistry.io/nested/namespace/image-name` is provided. It looks like it is having trouble parsing nested container repositories

### Expected Behavior

It correctly identifies nested image repositories.

### Steps To Reproduce

_No response_

### Environment

```markdown
- OS: macos
- updatecli: 0.33.2
```

### Anything else?

Related to my comment in https://github.com/updatecli/updatecli/issues/871#issuecomment-1250135152 I believe it will help a lot to switch to ggcr to parse/deal with docker images.

### PR DESCRIPTION
**Pull request by samj1912**
Fix #871 
Fix #872
Fix #606 

This PR guts the custom docker image/digest logic and replaces it with GGCR. This allows us to have a robust docker validation logic.

## Test

To test this pull request, you can run the following commands:

```yaml
sources:
  tag:
    kind: dockerimage
    spec:
      image: ghcr.io/updatecli/updatecli
  digest:
    kind: dockerdigest
    spec:
      image: ghcr.io/updatecli/updatecli
      tag: v0.34.0
```

```shell
./dist/updatecli_darwin_arm64/updatecli diff --config config.yaml


+++++++++++
+ PREPARE +
+++++++++++

Loading Pipeline "config.yaml"

SCM repository retrieved: 0


++++++++++++++++++
+ AUTO DISCOVERY +
++++++++++++++++++



++++++++++++
+ PIPELINE +
++++++++++++



###############
# CONFIG.YAML #
###############


SOURCES
=======

digest
------
✔ Docker Image Tag ghcr.io/updatecli/updatecli:v0.34.0 resolved to digest ghcr.io/updatecli/updatecli@sha256:779abc2c7a46c0cc0d568bac6e6d8b81c9b3255bb025a5f421d0911a82e3d4c5

tag
---
Searching for version matching pattern "latest"
✔ Docker Image Tag "v0.34.0" found matching pattern "latest"

=============================

REPORTS:


- CONFIG.YAML:
	Source:
		✔ [digest]  (kind: dockerdigest)
		✔ [tag]  (kind: dockerimage)

Run Summary
===========
Pipeline(s) run:
  * Changed:	0
  * Failed:	0
  * Skipped:	1
  * Succeeded:	0
  * Total:	1
```

### CODE REVIEW

File: pkg/plugins/resources/dockerdigest/main.go [hunk 1]
```diff
@@ -42,57 +35,30 @@ func New(spec interface{}) (*DockerDigest, error) {
...
-	if len(ds.spec.Username) > 0 && len(ds.spec.Password) == 0 {
-		validationErrors = append(validationErrors, "Docker registry username provided but not the password")
-	} else if len(ds.spec.Username) == 0 && len(ds.spec.Password) > 0 {
-		validationErrors = append(validationErrors, "Docker registry password provided but not the username")
-	}
-
-	// Return all the validation errors if found any
-	if len(validationErrors) > 0 {
-		return fmt.Errorf("validation error: the provided manifest configuration had the following validation errors:\n%s", strings.Join(validationErrors, "\n\n"))
+	keychains = append(keychains, authn.DefaultKeychain)
```
Comment (olblak) [inline pkg/plugins/resources/dockerdigest/main.go:53]: I like very much the improvement to use the local docker config file

**Comment by olblak** [comment]
Nice work, I don't realize how far you went. Indeed I need some e2e tests for different registries like ghcr.io dockerhub and quay

**Comment by olblak** [comment]
I did some initial testing and I didn't find any regression, I need to find some ways to test with other registries than dockerhub or ghcr.io

**Comment by olblak** [comment]
I confirm that your code fallback to the docker configuration file to retrieve credentials. 
I don't have permission to push on your local branch, so I'll probably move forward with your pullrequest and the e2e tests that I have locally

**Comment by samj1912** [comment]
The only 'regression' might be the registry for how dockerhub is resolved. It will be resolved to index.docker.io

File: pkg/plugins/resources/dockerdigest/main.go (hunk 1 above)
Comment (olblak) [reply pkg/plugins/resources/dockerdigest/main.go:53]: Which by the way work perfectly

File: pkg/plugins/resources/dockerdigest/main.go [hunk 2]
```diff
@@ -42,57 +35,30 @@ func New(spec interface{}) (*DockerDigest, error) {
...
+	arch := newSpec.Architecture
+	if arch == "" {
+		arch = "amd64"
 	}
+	newResource.options = append(newResource.options, remote.WithPlatform(v1.Platform{Architecture: arch, OS: "linux"}))
```
[the remark this answers is not in the document]
Comment (olblak) [reply pkg/plugins/resources/dockerdigest/main.go:58]: I'll leave this improvement for a later interation if someone ever need this

**Comment by olblak** [comment]
> The only 'regression' might be the registry for how dockerhub is resolved. It will be resolved to index.docker.io

Which I don't think is a problem at all. If I recall correctly that URL was needed for the previous code which is now replaced by GGCR

**Review by olblak — DISMISSED** [review]

**Comment by samj1912** [comment]
@olblak fixed the failing registry test and rebased. Should be all good to go now.

**Review by olblak — APPROVED** [review]

**Pull request merged**